DataFormats/ForwardDetId/interface/HGCSiliconDetId.h

```diff
@@ -0,0 +1,119 @@
+#ifndef DataFormats_ForwardDetId_HGCSiliconDetId_H
+#define DataFormats_ForwardDetId_HGCSiliconDetId_H 1
+
+#include <iosfwd>
+#include "DataFormats/DetId/interface/DetId.h"
+#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+
+class HGCSiliconDetId : public DetId {
+
+public:
+
+  enum waferType {HGCalFine=0, HGCalCoarseThin=1, HGCalCoarseThick=2};
+  static const int HGCalFineN  =12;
+  static const int HGCalCoarseN=8;
+  static const int HGCalFineTrigger  =3;
+  static const int HGCalCoarseTrigger=2;
+
+  /** Create a null cellid*/
+  HGCSiliconDetId();
+  /** Create cellid from raw id (0=invalid tower id) */
+  HGCSiliconDetId(uint32_t rawid);
+  /** Constructor from subdetector, zplus, layer, module, cell numbers */
+  HGCSiliconDetId(DetId::Detector det, int zp, int type, int layer, int waferU,
+	     int waferV, int cellU, int cellV);
+  /** Constructor from a generic cell id */
+  HGCSiliconDetId(const DetId& id);
+  /** Assignment from a generic cell id */
+  HGCSiliconDetId& operator=(const DetId& id);
+  
+  /** Converter for a geometry cell id */
+  HGCSiliconDetId geometryCell () const {return HGCSiliconDetId (det(), zside(), type(), layer(), waferU(), waferV(), 0, 0);}
+
+  /// get the subdetector
+  DetId::Detector subdet() const { return det(); }
+
+  /// get the type
+  int type() const { return (id_>>kHGCalTypeOffset)&kHGCalTypeMask; }
+
+  /// get the z-side of the cell (1/-1)
+  int zside() const { return (((id_>>kHGCalZsideOffset) & kHGCalZsideMask) ? -1 : 1); }
+
+  /// get the layer #
+  int layer() const { return (id_>>kHGCalLayerOffset)&kHGCalLayerMask; }
+
+  /// get the cell #'s in u,v or in x,y
+  int cellU() const { return (id_>>kHGCalCellUOffset)&kHGCalCellUMask; }
+  int cellV() const { return (id_>>kHGCalCellVOffset)&kHGCalCellVMask; }
+  std::pair<int,int> cellUV() const { return std::pair<int,int>(cellU(),cellV()); }
+  int cellX() const {
+    int N = (type() == HGCalFine) ? HGCalFineN : HGCalCoarseN;
+    return (3*(cellV()-N)+2);
```

Comment:
are these conversion formulas guaranteed to be the same from now until the end of the CMS detector? in other words, are we ever going to hit backward incompatibility issues because the geometry changes? (I'm thinking that maybe we should try to reserve some bits for a "version"...)

Reply:
I do not know - we do not have any spare bit. We shall simply grabage this and have a new definition

Reply:
This is a conversion formula from u-v to x-y axis as long as the axis definition and origins are the same - it does not matter for N so if we have even larger wafers and N=4 the formula will remain the same as long as we keep the axis definition. This one matches the one used by Ron & Zoltan in module design